libutils/file_lib.c

```diff
@@ -665,6 +672,13 @@ int safe_chown(const char *path, uid_t owner, gid_t group)
     int fd = safe_open(path, 0);
     if (fd < 0)
     {
+        // Do additional checking for FIFOs
+        struct stat st;
+        if (stat(path, &st) == 0 && S_ISFIFO(st.st_mode))
+        {
+            return chown(path, owner, group);
+        }
```

Comment:
So conceptually something like (not correct syntax/API, but you get the idea):

<pre>
fd = safe_open(dirname_path, O_RDONLY);
fstatat(fd, basename_path, &stat, AT_SYMLINK_NOFOLLOW);
gid = getgrnam("nogroup");
if (!gid) {
    gid = getgrnam("nobody"); // One of those should exist on all systems.
}
if (!gid) {
    return -1;
}
seteuid(stat.st_uid);
setegid(gid);
fchmodat(fd, basename_path, 0755)
seteuid(0); // Actually should save the old value.
setegid(0);
</pre>